File: Go/ParkingLot2-go/slot.go
package main

import "fmt"

type Slot2 struct {
	ID          int
	Type        string
	Vehicle     *Vehicle
	IsAvailable bool
}

func NewSlot2(id int, slotType string) (*Slot2, error) {
	if id <= 0 {
		return nil, fmt.Errorf("invalid slot!")
	}
	return &Slot2{ID: id, Type: slotType, IsAvailable: true}, nil
}

func (s *Slot2) Park(vehicle *Vehicle) bool {
	if !s.IsAvailable {
		fmt.Println("Slot is not empty!")
		return false
	}
	s.Vehicle = vehicle
	s.IsAvailable = false
	return true
}

func (s *Slot2) Unpark() {
	s.Vehicle = nil
	s.IsAvailable = true
}

File: JavaScript/ParkingLot2/Slot.js
class Slot {
  id = 0;
  type = '';
  vehicle = null;
  isAvailable = true;

  constructor(id, type) {
    if (!id || typeof id !== 'number' || id <= 0) {
      throw new Error('Invalid slot!');
    }

    this.id = id;
    this.type = type;
  }

  park(vehicle) {
    if (!this.isAvailable) {
      console.log('Slot is not empty!');
      return false;
    }

    this.vehicle = vehicle;
    this.isAvailable = false;

    return true;
  }

  unpark() {
    this.vehicle = null;
    this.isAvailable = true;
  }
}

module.exports = { Slot };
